R/pipeline.R

```diff
@@ -0,0 +1,221 @@
+#' outputs to a blob storage container is planned.
+#'
+#' @param config_path A string specifying the file path to the JSON
+#' configuration file. The configuration file should include at least `job_id`
+#' and `task_id` to manage the logging and task metadata.
+#' @param blob_storage_container Optional. The name of the blob storage
+#' container to which logs and outputs will be uploaded. If NULL, no upload
+#' will occur. (Planned feature, not currently implemented)
+#' @param output_dir A string specifying the directory where output, logs, and
+#' other pipeline artifacts will be saved. Defaults to the root directory ("/").
+#'
+#' @details
+#' The function reads the configuration from a JSON file and uses this to set
+#' up the job and task identifiers. It creates an output directory structure
+#' based on these IDs and starts logging the process in a file. The main
+#' pipeline process is handled by `process_pipeline()`, with errors caught and
+#' logged as warnings. The function will log the success or failure of the run.
+#'
+#' Logs are written to a file in the output directory, and console output is
+#' also mirrored in this log file. Error handling is in place to capture any
+#' issues during the pipeline execution and ensure they are logged
+#' appropriately.
+#'
+#' During the execution of the pipeline, the following output files are
+#' expected to be generated:
+#'
+#' - **Model Output**: An RDS file of the fitted model is saved in the
+#' task-specific directory (`model.rds`).
+#' - **Samples**: Parquet files containing the model's sample outputs are saved
+#' in a `samples` subdirectory, named using the `task_id` (e.g.,
+#' `task_id.parquet`).
+#' - **Summaries**: Parquet files summarizing the model's results are saved in
+#' a `summaries` subdirectory, also named using the `task_id` (e.g.,
+#' `task_id.parquet`).
+#' - **Logs**: A `logs.txt` file is generated in the task directory, capturing
+#' both console and error messages.
+#'
+#' The output directory structure will follow this format:
+#' ```
+#' <output_dir>/
+#' └── <job_id>/
+#'     ├── samples/
+#'     │   └── <task_id>.parquet
+#'     ├── summaries/
+#'     │   └── <task_id>.parquet
+#'     └── tasks/
+#'         └── <task_id>/
+#'             ├── model.rds
+#'             └── logs.txt
+#' ```
+#'
+#' @return The function does not return a value directly. However, the output
+#' directory will contain the following files:
+#' - Model RDS file (`model.rds`)
+#' - Sample output in Parquet format (`<task_id>.parquet` in the `samples/`
+#' directory)
+#' - Summary output in Parquet format (`<task_id>.parquet` in the `summaries/`
+#' directory)
+#' - Log file (`logs.txt`) in the task directory
+#'
+#' @rdname pipeline
+#' @export
+run_pipeline <- function(config_path,
+                         blob_storage_container = NULL,
+                         output_dir = "/") {
+  # TODO: Add config reader here
```

Comment:
Is there a config validator?

Reply:
Is "config reader" similar to "config validator"? As in "add config reader" would also include validation?

Reply:
See discussion here: https://github.com/CDCgov/cfa-epinow2-pipeline/issues/78

Reply:
I think the idea for the config validator repo is that before the [config generator](https://github.com/CDCgov/cfa-config-generator/) uploads them to blob, it sends them to the validator.

Maybe eventually we'll also swap from reading the JSON config into a list, to reading the JSON config into a statically defined object, which would simplify run time worry a bit.